modules/react/common/spec/components.test-d.tsx

```diff
@@ -0,0 +1,51 @@
+import React from 'react';
+import {expectTypeOf} from 'expect-type';
+
+import {createComponent, ElementComponent} from '../lib/utils/components';
+
+it('should assign an element-base component as an ElementComponent', () => {
+  const component = createComponent('div')({Component: (props: {foo: 'bar'}) => null});
+  expectTypeOf(component).toEqualTypeOf<ElementComponent<'div', {foo: 'bar'}>>();
+});
+
+it('should add sub-components to the signature', () => {
+  const component = createComponent('div')({
+    Component: (props: {foo: 'bar'}) => null,
+    subComponents: {
+      Foo: 'bar',
+    },
+  });
+  expectTypeOf(component).toEqualTypeOf<ElementComponent<'div', {foo: 'bar'}> & {Foo: number}>();
+});
+
+it('should assign ref and Element correctly for element components', () => {
+  createComponent('div')({
+    Component: (props: {}, ref, Element) => {
+      expectTypeOf(ref).toEqualTypeOf<React.Ref<HTMLDivElement>>();
+      expectTypeOf(Element).toEqualTypeOf<'div'>();
+      return null;
+    },
+  });
+});
+
+it('should assign ref and Element correctly for createComponent components', () => {
+  const component = createComponent('article')({Component: (props: {}) => null});
+
+  createComponent(component)({
+    Component: (props: {}, ref, Element) => {
+      expectTypeOf(ref).toEqualTypeOf<React.Ref<HTMLElement>>();
+      expectTypeOf(Element).toEqualTypeOf<ElementComponent<'article', {}>>();
+      return null;
+    },
+  });
+});
+
+it('should allow a valid ref when wrapping components', () => {
+  const Component = createComponent('button')({Component: (props: {}) => null});
+
+  createComponent('button')({
+    Component: (props: {}, ref, Element) => {
+      return <Component ref={ref} />;
```

Comment:
This is where the issue was. According to `PropsWithAs`, `ElementType` was a Component Class for Components created with `createComponent`. Passing a ref of an element caused signatures to not match. I had to extract the element type in `PropsWithAs` to fix this.

Reply:
Nit: This one seems a bit out of place without context since there's no `expect` call. Do you mind adding a comment referencing this PR or something?

Reply:
Hmm. I think this is more complicated than it needs to be. We really are ensuring `ref` can be sent to `<Component>`. It doesn't need another `createComponent` wrapper.

Reply:
I'm cleaning it up.